CI: Deploy Docker Images on remote VM

FabianUntermoser opened the pull request:
<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 👷️ Create small PRs.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

## Description
This PR sets up rudimentary deployment of the development branch to a remote VM `preview.vim.org`.

**Changes**:
- Adds second docker compose file specifying images that were pushed to the Github Container Registry
- Adds Github deploy action to deploy the app on a remote VM
  - this action should be triggered after either of the build jobs finish. However this does not seem to work currently. A guess is that the github actions need to be placed in the default branch for it to work. 
- Nginx configuration has been adjusted, config files were checked into the repo manually
- Changes to the VM have been documented in its respective changelog file
- Configures CMS Strapi backend to run behind the context_path defined in CMS_URL 
  - only when not in local development
  - for the development environment `https://preview.vim.org/cms/admin/` is used in the repository var CMS_URL
  - note that the CMS_URL variable is required to be set at build time in order for internal Strapi routes to properly use the context path correctly
- Adds cronjob to clean up stale docker images on the remote VM

Not included:
- Setup Database Migrations if necessary.
- Seems that the Content Manager for single types in Strapi is not reachable
  - `https://preview.vim.org/cms/admin/content-manager/singleType/api::home.home` not reachable
  - `https://preview.vim.org/cms/admin/content-manager/collectionType/api::newspost.newspost` works

## What type of PR is this? (check all applicable)
- [ ] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🤖 Build
- [x] 🔁 CI
- [ ] 📦 Chore
- [ ] ⏩ Revert

FabianUntermoser committed:
feat(ci): init deployment workflow

FabianUntermoser committed:
feat(ci): enable action for push on dev branch

FabianUntermoser committed:
fix(ci): use correct branch name

FabianUntermoser committed:
fix(ci): update github env secret names

FabianUntermoser committed:
feat(ci): test remote ssh command

jomunker committed:
refactor(ci): rename dev docker compose file to docker-compose.local.yml

jomunker committed:
feat(ci): add dev docker compose file

FabianUntermoser committed:
feat(ci): try to login to ghcrio container reg remotely

FabianUntermoser committed:
fix(ci): change secret var name

FabianUntermoser committed:
feat(ci): pull latest docker images remotely

FabianUntermoser committed:
feat(ci): deploy container images on remote vm

- prepare & copy .env file to server
- pull and start container images

FabianUntermoser committed:
fix(ci): fix remote working dir

FabianUntermoser committed:
feat(ci): stop remote script execution on failure

FabianUntermoser committed:
fix(ci): fix remote cd cmd

FabianUntermoser committed:
fix(ci): remove redundant docker .env file references

FabianUntermoser committed:
refactor(ci): ensure container startup

- pull,stop,start all containers
- show docker logs

FabianUntermoser committed:
fix(ci): use non-standard db port

FabianUntermoser committed:
fix(compose): use DATABASE_PORT variable

ensure configuring DATABASE_PORT in the .env file
will reflect in the db port being changed
as well as other services using the correct port

FabianUntermoser committed:
refactor(ci): update action name

FabianUntermoser committed:
feat(ci): deploy based on file changes

FabianUntermoser committed:
refactor(ci): ensure containers are recreated upon deployment

jomunker committed:
feat(ci): remove build context from web

jomunker committed:
feat(ci): remove unnecessary volumes

jomunker committed:
feat(ci): adjust cms secrets

jomunker committed:
feat(ci): update .env.example

FabianUntermoser committed:
feat(ci): scp database files to VM

jomunker committed:
fix(ci): web crashing because auth secret was missing in production mode

FabianUntermoser committed:
feat(ci): deploy branch docker image tag

jomunker committed:
feat(ci): change database port back to 3306

jomunker committed:
feat(ci): add feature/deployment-vm to on push trigger for cms build push pipeline

FabianUntermoser committed:
fix(ci): revert docker image tag override

jomunker committed:
feat(cms): add cms prefix to url on non local envs

FabianUntermoser committed:
fix(ci): pull branch image name

FabianUntermoser committed:
fix(ci): avoid copying db files

FabianUntermoser committed:
feat(cms): add URL variable

FabianUntermoser committed:
feat(cms): update cors to allow preview.vim.org

jomunker committed:
feat(cms): add admin flags to disable promotional features

jomunker committed:
feat: add environment variables to set telemetry and update flags of cms

jomunker committed:
feat(cms): hardcode url

FabianUntermoser committed:
fix(cms): avoid setting strapi url  on NODE_ENV=development

FabianUntermoser committed:
refactor: remove unnecessary compose version

[skip ci]

FabianUntermoser committed:
feat(cms): set CMS_URL at buildtime

- add Dockerfile arg
- set arg in github action through secret

FabianUntermoser committed:
fix(cms): ensure CMS_URL env var is used

FabianUntermoser committed:
feat(ci): print CMS_URL in CMS prod build

FabianUntermoser committed:
fix(ci): convert CMS_URL to env var instead of secret

FabianUntermoser committed:
test(ci): print env var

FabianUntermoser committed:
fix(ci): access CMS_URL as repo var

FabianUntermoser committed:
feat(ci): run deployment after images were pushed

FabianUntermoser committed:
feat(ci): allow manual run of deploy

FabianUntermoser committed:
refactor(ci): remove workflow_dispatch

FabianUntermoser committed:
refactor(ci): clean up log statement

FabianUntermoser committed:
refactor(ci): remove branch rule on workflow_run

FabianUntermoser committed:
feat(ci): deploy on dev branches

FabianUntermoser committed:
refactor(ci): avoid quotes in workflow_run

FabianUntermoser committed:
refactor(ci): remove temporary branch rules

jomunker committed:
chore(ci): add nginx configuration files

jomunker commented on .env.example:
```diff
@@ -1,15 +1,23 @@
+NEXTAUTH_SECRET=testsecret
 
+CMS_DATABASE_NAME=vim_cms
+CMS_URL="https://preview.vim.org/cms"
```
Can we delete this?

FabianUntermoser replied:
True we can, since local development does not set the CMS_URL

jomunker commented on .github/workflows/deploy-vm.yml:
```diff
@@ -0,0 +1,51 @@
+          cd ~/${{ github.event.repository.name }}/
+          echo ${{ secrets.GITHUB_TOKEN }} | docker login ghcr.io -u ${{github.actor}} --password-stdin
+          docker compose -f docker-compose.dev.yml pull
+          TAG=feature-deployment-vm docker compose -f docker-compose.dev.yml pull cms
```
Do we have to change this tag?

FabianUntermoser replied:
Good catch. I'll remove it.

jomunker commented on docker-compose.dev.yml:
```diff
@@ -0,0 +1,65 @@
+services:
+  web:
+    container_name: web
+    image: ghcr.io/vim/website_next_generation/web:${TAG:-development}
```
Does this work with the development branch?

FabianUntermoser replied:
Yes, if TAG is not set, it will resolve to the default value of `development`.

jomunker replied:
Okay so the `-` as prefix won't be recognized?

FabianUntermoser replied:
Yes the dash indicates the default value, its just bash syntax

jomunker commented on docker-compose.dev.yml:
```diff
@@ -0,0 +1,65 @@
+
+  cms:
+    container_name: strapi
+    image: ghcr.io/vim/website_next_generation/cms:${TAG:-development}
```
Same?

jomunker commented on docker-compose.dev.yml:
```diff
@@ -0,0 +1,65 @@
+      - vim-cms-data:/var/lib/mysql
+      - ./db/docker-scripts:/docker-entrypoint-initdb.d
+    ports:
+      - ${DATABASE_PORT}:3306
```
Do we need this in a variable?
If so we could also use it twice like so:
```suggestion
      - ${DATABASE_PORT}:${DATABASE_PORT}
```

FabianUntermoser replied:
If the VM does not have anything running on 3306 anymore, we could remove the value and hardcode it.
But I wouldn't use the variable twice. The right hand side of the port is for specifying the internal container port.
Changing it would also require to configure mysql to run on a different port, which does not serve any purpose.
I'd leave it as is, as sometimes it might be helpful to be able to change the port with a single command like this

`DATABASE_PORT=3307 docker compose -f docker-composer.dev.yml up -d `

jomunker commented on docker-compose.local.yml:
```diff
@@ -29,7 +28,7 @@ services:
       - ./db/docker-scripts:/docker-entrypoint-initdb.d
     ports:
-      - 3306:3306
+      - ${DATABASE_PORT}:3306
```
here as well with the port:

```suggestion
      - ${DATABASE_PORT}:${DATABASE_PORT}
```

jomunker commented on cms/.env.example:
```diff
@@ -1,3 +1,7 @@
+STRAPI_DISABLE_UPDATE_NOTIFICATION=false
+STRAPI_TELEMETRY_DISABLED=true
+NODE_ENV=local
```
I guess change this to development

```suggestion
NODE_ENV=development
```

FabianUntermoser committed:
feat(build): change NODE_ENV=development

Co-authored-by: Johannes Munker <56400587+jomunker@users.noreply.github.com>

FabianUntermoser committed:
refactor(build): remove CMS_URL env var not needed at runtime

CMS_URL var configures the context path for the STRAPI CMS backend.
It is required at build time in order to make STRAPI internal routes work.
It is not needed in local development, as there no contextpath is required.

FabianUntermoser committed:
feat(ci): run actions for 'feature/*' branches

FabianUntermoser committed:
feat(ci): add temporary deploy rule

FabianUntermoser committed:
refactor(ci): remove temporary deployment rules
